src/cowboy_decompress_h.erl

```diff
@@ -0,0 +1,217 @@
+		ignore=Ignore0, ratio_limit=RatioLimit0}) ->
+	Ignore = maps:get(decompress_ignore, Opts, Ignore0),
+	RatioLimit = maps:get(decompress_ratio_limit, Opts, RatioLimit0),
+	{Commands, Next} = cowboy_stream:info(StreamID, Info, Next0),
+	fold(Commands, State#state{next=Next, ignore=Ignore, ratio_limit=RatioLimit});
+info(StreamID, Info, State=#state{next=Next0}) ->
+	{Commands, Next} = cowboy_stream:info(StreamID, Info, Next0),
+	fold(Commands, State#state{next=Next}).
+
+-spec terminate(cowboy_stream:streamid(), cowboy_stream:reason(), #state{}) -> any().
+terminate(StreamID, Reason, #state{next=Next, inflate=Z}) ->
+	case Z of
+		undefined -> ok;
+		_ -> zlib:close(Z)
+	end,
+	cowboy_stream:terminate(StreamID, Reason, Next).
+
+-spec early_error(cowboy_stream:streamid(), cowboy_stream:reason(),
+	cowboy_stream:partial_req(), Resp, cowboy:opts()) -> Resp
+	when Resp::cowboy_stream:resp_command().
+early_error(StreamID, Reason, PartialReq, Resp, Opts) ->
+	cowboy_stream:early_error(StreamID, Reason, PartialReq, Resp, Opts).
+
+%% Internal.
+
+check_req(Req) ->
+	try cowboy_req:parse_header(<<"content-encoding">>, Req) of
+		undefined ->
+			#state{compress=undefined};
+		Encodings ->
+			case [E || E=(<<"gzip">>) <- Encodings] of
+				[] ->
+					#state{compress=undefined};
+				_ ->
+					#state{compress=gzip}
+			end
+	catch
+		_:_ ->
+			#state{compress=undefined}
+	end.
+
+fold(Commands, State) ->
+	fold(Commands, State, []).
+
+fold([], State, Acc) ->
+	{lists:reverse(Acc), State};
+fold([{response, Status, Headers0, Body}|Tail], State=#state{ignore=false}, Acc) ->
+	Headers = add_accept_encoding(Headers0),
+	fold(Tail, State, [{response, Status, Headers, Body}|Acc]);
+fold([{headers, Status, Headers0} | Tail], State=#state{ignore=false}, Acc) ->
+	Headers = add_accept_encoding(Headers0),
+	fold(Tail, State, [{headers, Status, Headers}|Acc]);
+fold([Command|Tail], State, Acc) ->
+	fold(Tail, State, [Command|Acc]).
+
+add_accept_encoding(Headers=#{<<"accept-encoding">> := AcceptEncoding}) ->
+	try cow_http_hd:parse_accept_encoding(iolist_to_binary(AcceptEncoding)) of
+		List ->
+			case lists:keyfind(<<"gzip">>, 1, List) of
+				%% gzip is excluded but this handler is not ignored; we replace.
+				{_, 0} ->
+					Replaced = lists:keyreplace(<<"gzip">>, 1, List, {<<"gzip">>, 1000}),
+					Codings = build_accept_encoding(Replaced),
+					Headers#{<<"accept-encoding">> => Codings};
+				{_, _} ->
+					Headers;
+				false ->
+					case lists:keyfind(<<"*">>, 1, List) of
+						%% Others are excluded along with gzip; we add.
+						{_, 0} ->
+							WithGzip = [{<<"gzip">>, 1000} | List],
+							Codings = build_accept_encoding(WithGzip),
+							Headers#{<<"accept-encoding">> => Codings};
+						{_, _} ->
+							Headers;
+						false ->
+							Headers#{<<"accept-encoding">> => [AcceptEncoding, <<", gzip">>]}
+					end
+			end
+	catch _:_ ->
+		Headers#{<<"accept-encoding">> => <<"gzip">>}
```

Comment:
I don't know if that's the right approach, overriding the header completely. Maybe this is a case where we would be better off crashing: if you send a response with a broken accept-encoding header and use this stream handler, you deserve the crash IMO.

Reply:
I'm thinking that we shouldn't crash but we should not overwrite the accept-encoding header the user has set. If the user is doing something weird then they can include gzip themselves. This is very much an edge case anyway so let's go with that for now and see later if that causes problems.

Reply:
Hm but we already do replace in cowboy_compress_h when a header is invalid so I guess we can keep this same behavior here. No change needed here in that case.